### PR TITLE
Introduces LOCAL_COMPONENT("lc") binary annotation and example use-case

### LINKED ISSUE #808: Support local spans via a LOCAL_COMPONENT("lc") binary annotation
Currently it is tribal knowledge that local (in-process) spans exist. We need a way to designate span as local, and offer insight into how to use them. This can lead us to answer what status quo could be on this, and how they might be tested.

Firstly, local spans need to be annotated with a namespace that can be used like service. A discussion on github suggests using a LOCAL_COMPONENT("lc") binary annotation for this.

Here's a json view:

``` json
{
  "key" : "lc",
  "value" : "encryption-service",
  "endpoint" : {
    "serviceName" : "proxy",
    "ipv4" : "1.2.3.4",
    "port" : 345
  }
}
```

You'll notice that the binary annotation holds almost all information needed to identify the context of the span. It contains a searchable annotation which contains its namespace, in this case `encryption-service`. Via the existing contract of binary annotations, it also includes the service that contains it, in this case `proxy`.

What's left is the skeleton of the span. How do we capture the duration of the span?

I'm aware of three options:
- Re-use "cs/cr" annotations
- Make up new annotations for the bookends of a local span.
- piggy-back off Span.startTs/duration (#807)

TL;DR; is that I think the latter is best.

If we re-use "cs/cr" annotations, we avoid logic around dependency graphs where server annotations are special-cased. In other words, we avoid accidentally cluttering it with local spans. We can also hang off an existing explanation where we use "cs/cr" to address uninstrumented systems. It might be the least code changes.

``` json
{
  "traceId" : "0000000000000005",
  "name" : "encrypt",
  "id" : "000000000000029a",
  "parentId" : "0000000000000002",
  "annotations" : [
    {
      "timestamp" : 60,
      "value" : "cs",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    },
    {
      "timestamp" : 100,
      "value" : "cr",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    }
  ],
  "binaryAnnotations" : [
    {
      "key" : "lc",
      "value" : "encryption-service",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    }
  ]
}
```

If we make up new annotations for the bookends, we avoid adding more cognitive weight to "cs/cr", which are already confusing. Instead, all two-character annotation keys prefixed 'l' explain local spans, and those not interested simply don't read their docs. We do have to add more constants, and affect the instrumentation more. While this is nice for documentation containment, it could look overwhelming.

``` json
{
  "traceId" : "0000000000000005",
  "name" : "encrypt",
  "id" : "000000000000029a",
  "parentId" : "0000000000000002",
  "annotations" : [
    {
      "timestamp" : 60,
      "value" : "ls",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    },
    {
      "timestamp" : 100,
      "value" : "lr",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    }
  ],
  "binaryAnnotations" : [
    {
      "key" : "lc",
      "value" : "encryption-service",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    }
  ]
}
```

If we piggy-back off Span.startTs/duration (#807), local spans really only need span metadata and the single binary annotation. This would result in less storage and cognitive burden, as we no longer need to think about annotations in order to understand the duration of a span. This is the most dramatic departure from existing "core annotations" so needs to be specifically supported in UI and presentation.

``` json
{
  "traceId" : "0000000000000005",
  "name" : "encrypt",
  "id" : "000000000000029a",
  "parentId" : "0000000000000002",
  "startTs" : 60,
  "duration" : 40,
  "binaryAnnotations" : [
    {
      "key" : "lc",
      "value" : "encryption-service",
      "endpoint" : {
        "serviceName" : "proxy",
        "ipv4" : "1.2.3.4",
        "port" : 345
      }
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Local spans have been discussed at length, but have never been
formalized. This formalizes how to tag a span as local, including
rationale documentation. To help demonstrate the point, zipkin-web and
zipkin-query use this for bootstrap tracing.

Fixes #808

Here are example local spans from finatra and twitter-server bootstrap:

screen shot 2015-11-07 at 12 34 17 pm

screen shot 2015-11-07 at 12 35 28 pm

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
cc @yurishkuro @eirslett @spencergibb @dsyer @kristofa @michaelsembwever

**Comment by codefromthecrypt** [comment]
Note that finagle instrumentation doesn't yet support Span.duration. The workaround until that's not the case is add a ClientRecv("cr") annotation. That's because finagle needs some signal to know the span is complete. https://github.com/twitter/finagle/blob/develop/finagle-zipkin/src/main/scala/com/twitter/finagle/zipkin/thrift/DeadlineSpanMap.scala#L110

**Comment by yurishkuro** [comment]
Does this change require UI/JS changes? When I emit a span with LC only, without RPC-bounding annotations, the UI doesn't show it.

**Comment by codefromthecrypt** [comment]
Make sure you are setting Span.timestamp, duration? The example of
bootstrap tracing displays...

File: zipkin-query-service/src/main/scala/com/twitter/zipkin/builder/QueryServiceBuilder.scala [hunk 1]
```diff
@@ -31,23 +26,24 @@ case class QueryServiceBuilder(override val defaultFinatraHttpPort: String = "0.
...
-                                          Builder[RuntimeEnvironment => ListeningServer] {
+                                ) extends ZipkinQueryServer(spanStore, dependencies) {
 
-  override def apply() = (runtime: RuntimeEnvironment) => {
-    LoggerFactory.getLogger(org.slf4j.Logger.ROOT_LOGGER_NAME)
-      .asInstanceOf[Logger].setLevel(Level.toLevel(logLevel))
-
-    /** If the span transport is set, trace accordingly, or disable tracing */
+  /** If the span transport is set, trace accordingly, or disable tracing. */
+  def premain {
```
Comment (codefromthecrypt) [inline zipkin-query-service/src/main/scala/com/twitter/zipkin/builder/QueryServiceBuilder.scala]: @yurishkuro this was a bug!! maybe related to your inability to see spans. pushing a fix now.

**Comment by codefromthecrypt** [comment]
rebased with a fix to a bug where the tracers weren't initializing.

@yurishkuro @kristofa please take a look. I'd like to add support to this to brave, but want to make sure folks are on-board before raising that.

**Comment by codefromthecrypt** [comment]
last call for feedback, I'll merge in a couple hours and address anything else post-merge

**Comment by yurishkuro** [comment]
looks good

**Comment by yurishkuro** [comment]
@adriancole I pulled this, but still not able to see local spans, even though I do send ts and duration. Your screenshots above show "cr" for both examples of local spans. Did you try it without having any RPC annotations on the local span?

**Comment by codefromthecrypt** [comment]
good point. so we have a test that shows spanstore doesn't require annotations, but maybe there's some special casing in the web process subverting this. Should be possible to reproduce by making json and then posting it to the query server. I'll have a look.

**Comment by yurishkuro** [comment]
the json result from query service does not return that local span either, so it might be deeper than web layer

**Comment by nishantshobhit** [comment]
brave user here: The trace_id generate for a span generated from the local tracer is different from other spans for the same trace. may be that's the problem here?

**Comment by codefromthecrypt** [comment]
@nishantshobhit feel free to hop on openzipkin/zipkin even if the questions are about brave.

I have a working example of local tracer in the zipkin-java project. However, there's some important details about the implementation of span storage in brave you need to keep in mind. Look here for details: https://github.com/openzipkin/brave/blob/master/brave-core/src/main/java/com/github/kristofa/brave/LocalSpanState.java#L9
